### PR TITLE
Multiple pipelines

### PR DESCRIPTION
**Pull request by Neloop**

### CODE REVIEW

File: app/helpers/ExerciseConfig/Config/Environment.php [hunk 1]
```diff
@@ -31,15 +31,18 @@ public function getPipelines(): array {
 
   /**
    * Get pipeline for the given name.
```
Comment (krulis-martin) [inline app/helpers/ExerciseConfig/Config/Environment.php]: Get pipeline of the given ID.

File: app/helpers/ExerciseConfig/Helper.php [hunk 2]
```diff
@@ -19,133 +20,147 @@ class Helper {
...
+    $this->joinPipelinesAndGetInputVariables($pipelinesIds, $inputs, $references);
 
-    // initialize results array
+    // go through inputs and assign them to result
     $result = [];
-    foreach ($pipelines as $pipelineId => $pipeline) {
-      $result[$pipelineId] = [];
-    }
+    for ($i = 0; $i < count($pipelinesIds); $i++) {
+      $pipelineId = $pipelinesIds[$i];
```
Comment (krulis-martin) [inline app/helpers/ExerciseConfig/Helper.php]: foreach ($pipelinesIds as $i => $pipelineId)

File: app/helpers/ExerciseConfig/Helper.php [hunk 3]
```diff
@@ -185,76 +200,79 @@ public function getEnvironmentsForFiles(IExercise $exercise, array $files): arra
...
+              $variable = $pipeline->getVariablesTable()->get($varName);
+            }
 
-          // we only seek for the local file variables, not the remote ones...
-          // although port might be of type file, variables assigned to it
-          // may have remote-file type in case the file should be downloaded
-          if (!$variable->isFile()) {
-            continue;
-          }
+            // somethings fishy here
```
Comment (krulis-martin) [inline app/helpers/ExerciseConfig/Helper.php]: It is hardly a fault of a fish that a variable has not been set. Unless you blame my code in front-end and you are mixing up fishes and beavers as both have positive inclination towards water.

File: app/helpers/ExerciseConfig/Validation/ExerciseConfigValidator.php [hunk 4]
```diff
@@ -115,29 +108,28 @@ private function checkPipelinesSection(Exercise $exercise, array $pipelines,
...
-    $expectedVariables = $this->helper->getVariablesForExercise($pipelineConfigs, $environmentVariables);
+    $expectedVariables = $this->helper->getVariablesForExercise($pipelinesIds, $environmentVariables);
 
     /**
      * @var string $pipelineId
      * @var PipelineVars $pipelineVars
      */
-    foreach ($pipelines as $pipelineVars) {
+    for ($i = 0; $i < count($pipelines); $i++) {
+      $pipelineVars = $pipelines[$i];
```
Comment (krulis-martin) [inline app/helpers/ExerciseConfig/Validation/ExerciseConfigValidator.php]: Why, oh, why? Foreach could do the same job as a signle-liner.

File: app/helpers/ExerciseConfig/Helper.php [hunk 5]
```diff
@@ -19,133 +20,147 @@ class Helper {
...
+          // port is file and also array, in exercise config if there should be
+          // defined file as variable it is expected to be remote file, so the
+          // remote file type is offered back to web-app
+          $variable = (new Variable(VariableTypes::$REMOTE_FILE_ARRAY_TYPE))->setName($variableName);
+        } else if ($inputPort->isFile() && !$inputPort->isArray()) {
+          // port is file and not an array, in exercise config if there should be
+          // defined file as variable it is expected to be remote file, so the
+          // remote file type is offered back to web-app
+          $variable = (new Variable(VariableTypes::$REMOTE_FILE_TYPE))->setName($variableName);
+        } else {
```
[the remark this answers is not in the document]
Comment (Neloop) [reply app/helpers/ExerciseConfig/Helper.php]: Well, this was previously written code, which I did not modify for the sake of my sanity. Also in my opinion, this is more readable than ternary.

[the remark this answers is not in the document]
Comment (krulis-martin) [reply app/helpers/ExerciseConfig/Helper.php]: If three and half lines of four are identical in both branches, I cannot agree.

File: app/helpers/ExerciseConfig/Helper.php [hunk 6]
```diff
@@ -231,7 +225,7 @@ public function getEnvironmentsForFiles(IExercise $exercise, array $files): arra
               $variable = $pipeline->getVariablesTable()->get($varName);
             }
 
-            // somethings fishy here
+            // somethings beavery here
```
Comment (krulis-martin) [inline app/helpers/ExerciseConfig/Helper.php]: http://weknowmemes.com/2012/03/for-seal-thanks/

**Pull request merged**